### DISCUSSION TITLE
Too few arguments to function Shlinkio\\Shlink\\Config\\parseEnvVar(), 0 passed in vendor/shlinkio/shlink-config/functions/functions.php on line 82 and exactly 1 expected invendor/shlinkio/shlink-config/functions/functions.php:46

### DISCUSSION THREAD

**Discussion by Emojigit** [shlinkio/discussions#127] [Help wanted] [answered]
#### How Shlink is set up

* Shlink Version: https://github.com/shlinkio/shlink/commit/0e9ea5027caaa24be16344a0cefe6d226fb56bd8 (`3.5.2`)
* PHP Version: 8.1 (`8.1.2-1ubuntu2.11`)
* How do you serve Shlink: Self-hosted Apache
* Database engine used: MariaDB (Followed the installation instructions of MySQL, it should be compactible)

#### Summary
I installed Shlink using the following commands (just like what developers would, but on the stable branch):
```bash
git clone https://github.com/shlinkio/shlink/ -b main --depth 1 --recursive # Clone the stable branch, shalow clone, recursive submodules (if any)
composer install # install PHP dependencies
```

I set up Shlink using `vendor/bin/shlink-install`. Here is my `config/params/generated_config.php`:
```php
<?php
return array(
    'DB_DRIVER' => 'mysql',
    'DB_NAME' => 'shlink',
    'DB_HOST' => 'localhost',
    'DB_PORT' => '3306',
    'DB_USER' => 'shlink',
    'DB_PASSWORD' => '[PLACEHOLDER]',
    'DB_UNIX_SOCKET' => '/run/mysqld/mysqld.sock',
    'DEFAULT_DOMAIN' => 'placeholder.example.com',
    'IS_HTTPS_ENABLED' => true,
    'DEFAULT_SHORT_CODES_LENGTH' => 5,
    'AUTO_RESOLVE_TITLES' => true,
    'REDIRECT_APPEND_EXTRA_PATH' => false,
    'MULTI_SEGMENT_SLUGS_ENABLED' => false,
    'SHORT_URL_TRAILING_SLASH' => true,
    'SHORT_URL_MODE' => 'strict',
    'GEOLITE_LICENSE_KEY' => null,
    'REDIRECT_STATUS_CODE' => 308,
    'REDIRECT_CACHE_LIFETIME' => 30,
    'DISABLE_TRACKING' => false,
    'TRACK_ORPHAN_VISITS' => true,
    'DISABLE_TRACK_PARAM' => 'dnt',
    'DISABLE_TRACKING_FROM' => '127.0.0.1,::1',
    'DISABLE_IP_TRACKING' => true,
    'DISABLE_UA_TRACKING' => true,
    'DISABLE_REFERRER_TRACKING' => true,
    'DEFAULT_BASE_URL_REDIRECT' => null,
    'DEFAULT_INVALID_SHORT_URL_REDIRECT' => null,
    'DEFAULT_REGULAR_404_REDIRECT' => null,
    'DEFAULT_QR_CODE_SIZE' => 300,
    'DEFAULT_QR_CODE_MARGIN' => 0,
    'DEFAULT_QR_CODE_FORMAT' => 'svg',
    'DEFAULT_QR_CODE_ERROR_CORRECTION' => 'q',
    'DEFAULT_QR_CODE_ROUND_BLOCK_SIZE' => true,
    'DELETE_SHORT_URL_THRESHOLD' => null,
    'BASE_PATH' => '',
    'TIMEZONE' => null,
    'REDIS_SERVERS' => null,
);
```

Here I attach my Apache site configuration file:
```
<IfModule mod_ssl.c>
<VirtualHost *:443>
    ServerName placeholder.example.com

    DocumentRoot /var/www/s/public

    ErrorLog ${APACHE_LOG_DIR}/error-s.log
    CustomLog ${APACHE_LOG_DIR}/access-s.log combined

    # SSL certs and configurations are configured here; 
    # The server is proxied through Cloudflare.
    Include /etc/apache2/cf_certs/cf_certs.conf 
    <Directory "/var/www/s/public">
        Options FollowSymLinks Includes ExecCGI
        AllowOverride all
        Order allow,deny
        Allow from all
    </Directory>

    Header set Access-Control-Allow-Origin "app.shlink.io"
</VirtualHost>
</IfModule>
```
#### Current behavior
The PHP script failed to execute. It raises the following error:
```
PHP Fatal error:  Uncaught ArgumentCountError: Too few arguments to function Shlinkio\\Shlink\\Config\\parseEnvVar(), 0 passed in /var/www/s/vendor/shlinkio/shlink-config/functions/functions.php on line 82 and exactly 1 expected in /var/www/s/vendor/shlinkio/shlink-config/functions/functions.php:46
Stack trace:
#0 /var/www/s/vendor/shlinkio/shlink-config/functions/functions.php(82): Shlinkio\\Shlink\\Config\\parseEnvVar()
#1 /var/www/s/config/autoload/swoole.global.php(26): Shlinkio\\Shlink\\Config\\getOpenswooleConfigFromEnv()
#2 /var/www/s/config/autoload/swoole.global.php(34): Laminas\\ConfigAggregator\\PhpFileProvider::{closure}()
#3 /var/www/s/vendor/laminas/laminas-config-aggregator/src/PhpFileProvider.php(29): include('...')
#4 /var/www/s/vendor/laminas/laminas-config-aggregator/src/ConfigAggregator.php(238): Laminas\\ConfigAggregator\\PhpFileProvider->__invoke()
#5 /var/www/s/vendor/laminas/laminas-config-aggregator/src/ConfigAggregator.php(82): Laminas\\ConfigAggregator\\ConfigAggregator->loadConfigFromProviders()
#6 /var/www/s/config/config.php(52): Laminas\\ConfigAggregator\\ConfigAggregator->__construct()
#7 /var/www/s/config/container.php(26): require('...')
#8 /var/www/s/config/container.php(31): {closure}()
#9 /var/www/s/config/run.php(10): include('...')
#10 /var/www/s/public/index.php(5): {closure}()
#11 {main}
  thrown in /var/www/s/vendor/shlinkio/shlink-config/functions/functions.php on line 46
```

#### Expected behavior
It should work. I've followed the steps in [install from dist files](https://shlink.io/documentation/install-dist-file/).

#### How to reproduce
Follow what I did. If this is not a device-specific problem nor a careless mistake, this should be reproduced.

**Comment by acelaya** [comment]
In the document you reference it doesn't mention anything about git-cloning. Instead, you should download a "dist" file.

In that same document you will find a "Download" button. Click on it, and on the release page, download the appropriate dist file, decompress it, and then follow the next steps as you did.

**Reply by Emojigit** [reply]
The same error still raises after migrating to dist files. I've coped `config/params/generated_config.php` from the old one to the new (dist file) one.

**Reply by Emojigit** [reply]
My cloned git version and the released dist version shares the same git hash, so the config files should be compactible.

**Reply by acelaya** [reply]
It's not just the git hash. There's a building process around generating the dist files. They're are not just a zip containing the same you clone from sources.

**Reply by Emojigit** [reply]
I've properly installed the dist files. I've observed a strange behaviour: Web accesses are always 500, but cli access is normal (i.e. I can create new short URLs without error reported).

**Reply by acelaya** [reply]
Try deleting the `data/cache/app_config.php` file (if exists).

If that doesn't help, check the logs inside `data/log`, and paste here waht you see there.

If there are no logs there, try checking apache logs, but in that case, it is very likely the error is due to missing permissions (third setp from the installation instructions: `Recursively grant write permissions to the "data" directory.`).

**Reply by Emojigit** [reply]
The file `data/cache/app_config.php` does not exist.

This is the only log:

```
[2023-03-11T15:57:39.005532+08:00] [7425dfe6-cb3c-4440-8e4a-516d493f2168] Shlink.WARNING - Tried to locate visit with id "2", but a GeoLite2 db was not found.
```

I have never provided a GeoLite token.

**Reply by acelaya** [reply]
And what about Apache logs?

**Reply by acelaya** [reply, marked as answer]
@Emojigit FYI, another user faced this same issue and everything points in the direction of it being a bug on PHP 8.1.2 https://github.com/shlinkio/shlink/discussions/1721#discussioncomment-5343072

Try updating to the latest PHP 8.1 patch version.

**Reply by Emojigit** [reply]
OK, let me see is it possible to (and which package to) upgrade on Ubuntu...

**Reply by thomasgg23** [reply]
@Emojigit did you find a solution to fix/update the apache package on an "vanilla" ubuntu?
The only way I know is to install the sury repository and see if the bug is fixed there.
https://deb.sury.org/

**Reply by thomasgg23** [reply]
PHP 8.1 on ubuntu with the sury repository still not working, have to switch to PHP 8.2 in the sury repository to get shlink working on ubuntu 22.04 LTS
